libraries/chain/controller.cpp

```diff
@@ -2592,20 +2592,51 @@ block_state_ptr controller::finalize_block( const signer_callback_type& signer_c
 
    auto& ab = std::get<assembled_block>(my->pending->_block_stage);
 
+   auto pfa = *ab._pending_block_header_state.prev_activated_protocol_features.get();
+   auto pfs = my->protocol_features.get_protocol_feature_set();
+
    auto bsp = std::make_shared<block_state>(
                   std::move( ab._pending_block_header_state ),
                   std::move( ab._unsigned_block ),
                   std::move( ab._trx_metas ),
-                  my->protocol_features.get_protocol_feature_set(),
+                  pfs,
                   []( block_timestamp_type timestamp,
                       const flat_set<digest_type>& cur_features,
                       const vector<digest_type>& new_features )
-                  {},
-                  signer_callback
+                  {}
               );
 
    my->pending->_block_stage = completed_block{ bsp };
 
+   if( my->block_sign_fut.valid() ) {
+      auto [pfa_preserved, pfs_preserved, result] = my->block_sign_fut.get();
+
+      if (!result->additional_signatures.empty()) {
+         bool wtmsig_enabled = detail::is_builtin_activated(pfa_preserved, pfs_preserved, builtin_protocol_feature_t::wtmsig_block_signatures);
+
+         constexpr auto additional_sigs_eid = additional_block_signatures_extension::extension_id();
+
+         EOS_ASSERT(wtmsig_enabled, block_validate_exception,
+                    "Block has multiple signatures before activation of WTMsig Block Signatures");
+
+         // as an optimization we don't copy this out into the legitimate extension structure as it serializes
+         // the same way as the vector of signatures
+         static_assert(fc::reflector<additional_block_signatures_extension>::total_member_count == 1);
+         static_assert(std::is_same_v<decltype(additional_block_signatures_extension::signatures), std::vector<signature_type>>);
+
+         emplace_extension(result->block->block_extensions, additional_sigs_eid, fc::raw::pack( result->additional_signatures ));
+      }
+
+      my->emit( accepted_block, result );
+      my->log_irreversible();
+
+   }
+
+   my->block_sign_fut = async_thread_pool( my->block_sign_pool.get_executor(),
+                                           [pfa, pfs, bsp, signer_callback]() mutable {
+                                              (*bsp).sign( bsp->block, signer_callback );
```

Comment:
signer_callback should be moved into the lambda

Reply:
Done.